### ISSUE TITLE
Custom Pearson Correlation Loss Taking Unfeasible Values During Training

### ISSUE THREAD

**Issue by adamw96** [tensorflow/tensorflow#37495]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): **Windows 10**
- TensorFlow installed from (source or
binary): **Source**
- TensorFlow version (use command below): **TensorFlow 2.1.0**
- Python version: **Python 3.6.10 (Anaconda Distribution)**
- CUDA/cuDNN version: **CUDA 10.1; cuDNN 7.6.5 (Nov 5, 2019)**
- GPU model and memory: **NVIDIA GeForce RTX 2070;  8GB VRAM**

**Describe the current behavior**
Loss values for a custom loss function that computes sample correlations takes on unfeasible values (i.e. 14-15) with certain model architectures; i.e. if hidden layers are too wide.

**Describe the expected behavior**
The correlation function's implementation should bound it to within [-1,1], and the custom loss function takes the negative absolute so it should be within [-1,0].

**Standalone code to reproduce the issue** 
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import tensorflow as tf
from tensorflow.keras import regularizers
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.initializers import he_normal
import tensorflow_probability as tfp

# Data
np.random.seed(42)
data_X = np.random.random_sample((100,6))
data_Y = np.random.random_sample((100,1))

# Feed-forward NN
model = Sequential()

# Input Layer
model.add(Dense(6, activation='linear', input_shape=(6,)))

# Hidden Layers: Arbitrarily wide
model.add(Dense(5000, 
                activation='relu', 
                kernel_initializer=he_normal(), 
                kernel_regularizer=regularizers.l1(10**-3)
                ))
model.add(Dropout(0.3))

# Output Layer; Regression
model.add(Dense(1, activation='sigmoid'))

# Custom Loss Function
def MaxCorrelation(y_true,y_pred):
    """
    Goal is to maximize correlation between y_pred, y_true. Same as minimizing the negative.
    """
    return -tf.math.abs(tfp.stats.correlation(y_pred,y_true, sample_axis=None, event_axis=None))

# Compilation
model.compile(
            optimizer='adam', 
            loss= MaxCorrelation,
)

# Train the model
history = model.fit(data_X, data_Y,
          epochs=5,
          batch_size = 32,
          verbose=1,
         )
```
**Other info / logs** 
```
OUT: Train on 100 samples
Epoch 1/5
100/100 [==============================] - 0s 2ms/sample - loss: 14.1344
Epoch 2/5
100/100 [==============================] - 0s 90us/sample - loss: 14.0880
Epoch 3/5
100/100 [==============================] - 0s 90us/sample - loss: 13.9630
```

Testing the loss function manually seems to work as intended:
```
import tensorflow.keras.backend as K
Y_Pred = model.predict(data_X).astype(float)
K.eval(MaxCorrelation(data_Y, Y_Pred))
--
OUT: -0.137205319813903
```

My understanding is that for each `batch_size=32`, the model returns 32 samples of scalar values from the output layer which is squashed between (0,1) and the loss MaxCorrelation is evaluated on the 32 predictions vs 32 targets. Thus the losses should still be within the bounds unless some aggregation is happening per epoch (i.e. one of the batch losses tending to infinity?), but even when removing the negative sign the loss is still positive. This issue only seems to persist when the hidden layers are wide/deep enough.

I have also tried replacing the loss function with some of the backend implementations [here](https://stackoverflow.com/questions/46619869/how-to-specify-the-correlation-coefficient-as-the-loss-function-in-keras?rq=1), but similar results.

**Comment by gadagashwini** [comment]
I could reproduce the issue with Tf 2.1 on colab.
Please take a look at gist [here](https://colab.sandbox.google.com/gist/gadagashwini/a44f5f895fb6c07638a27b78fe7db355/untitled437.ipynb). Thanks

**Comment by ravikyram** [comment]
I could reproduce the issue with TF 2.2-rc3 on colab.
Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/444e226ca2509e88fe293d9606f852b6/untitled845.ipynb).Thanks!

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/7d061c446be18466d7ccbdf091b13040/untitled7.ipynb). Thanks!

**Comment by Saduf2019** [comment]
@adamw96 
please refer to the [latest execution](https://colab.research.google.com/gist/Saduf2019/f4cf2ba9296f4e7e3376d7eb35cb0f1e/untitled638.ipynb) on tf nightly and confirm if the issue is resolved
